framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -2678,40 +2679,78 @@ protected void checkTypecastSafety(TypeCastTree typeCastTree) {
             }
             reported = true; // don't issue cast unsafe warning.
         }
-
         // Don't call TypeHierarchy#isSubtype(exprType, castType) because the underlying Java types
         // will not be in the correct subtyping relationship if this is a downcast.
-        if (!reported && !isTypeCastSafe(castType, exprType)) {
-            checker.reportWarning(
-                    typeCastTree, "cast.unsafe", exprType.toString(true), castType.toString(true));
+        if (!reported) {
+            TypecastKind castResult = isTypeCastSafe(castType, exprType);
+
+            if (castResult == TypecastKind.WARNING) {
+                checker.reportWarning(
+                        typeCastTree,
+                        "cast.unsafe",
+                        exprType.toString(true),
+                        castType.toString(true));
+            } else if (castResult == TypecastKind.ERROR) {
```

Comment:
Why does this only handle `WARNING` and `ERROR` and not also at least `NOT_UPCAST`?
It seems a bit odd to have the `TypecastKind` with five values if `isTypeCastSafe` only returns three of the options.
Can you think of cleaning up the logic and the enum to make this easier to follow? Maybe there should be two separate enums, one for `isTypeCastSafe` and one for `isUpcast`?